### PR TITLE
trayer: new service added

### PR DESCRIPTION
**Pull request by AndreasMager**
### Description
trayer is a lightweight GTK2+ systray for UNIX desktops

### Checklist

- [x] Change is backwards compatible.

- [x] Code formatted with `./format`.

- [x] Code tested through `nix-shell --pure tests -A run.all`.

- [x] Test cases updated/added. See [example](https://github.com/nix-community/home-manager/commit/f3fbb50b68df20da47f9b0def5607857fcc0d021#diff-b61a6d542f9036550ba9c401c80f00ef).

- [x] Commit messages are formatted like

    ```
    {component}: {description}

    {long description}
    ```

    See [CONTRIBUTING](https://github.com/nix-community/home-manager/blob/master/doc/contributing.adoc#sec-commit-style) for more information and [recent commit messages](https://github.com/nix-community/home-manager/commits/master) for examples.

- If this PR adds a new module

  - [x] Added myself as module maintainer. See [example](https://github.com/nix-community/home-manager/blob/068ff76a10e95820f886ac46957edcff4e44621d/modules/programs/lesspipe.nix#L6).

  - [x] Added myself and the module files to `.github/CODEOWNERS`.

### CODE REVIEW

File: modules/services/trayer.nix [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+
+  cfg = config.services.trayer;
+
+in {
+  meta.maintainers = [ maintainers.mager ];
+
+  options = {
+    services.trayer = {
+      enable =
+        mkEnableOption "trayer the lightweight GTK2+ systray for UNIX desktops";
```
Comment (sumnerevans) [inline modules/services/trayer.nix:15]: nit. As is, this will be rendered as:
> Whether or not to enable trayer the lightweight GTK2+ systray for UNIX desktops.
in the generated docs.

```suggestion
        mkEnableOption "trayer, the lightweight GTK2+ systray for UNIX desktops";
```

File: modules/services/trayer.nix [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+        description = ''
+          Trayer configuration as a set of attributes.
+        '';
+        default = { };
+        example = {
+          edge = "top";
+          padding = 6;
+          SetDockType = true;
+          tint = "0x282c34";
+        };
```
Comment (sumnerevans) [inline modules/services/trayer.nix:36]:
```suggestion
        example = literalExample ''
          {
            edge = "top";
            padding = 6;
            SetDockType = true;
            tint = "0x282c34";
          }
        '';
```

File: modules/services/trayer.nix [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+        example = literalExample "pkgs.trayer";
+        description = "The package to use for the trayer binary.";
+      };
+
+      config = mkOption {
+        type = with types; attrsOf (nullOr (either str (either bool int)));
+        description = ''
+          Trayer configuration as a set of attributes.
+        '';
+        default = { };
```
Comment (sumnerevans) [inline modules/services/trayer.nix:30]:
```suggestion
        default = { };
        defaultText = literalExample "{ }";
```

File: modules/services/trayer.nix [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+        defaultText = literalExample "pkgs.trayer";
+        type = types.package;
+        example = literalExample "pkgs.trayer";
+        description = "The package to use for the trayer binary.";
+      };
+
+      config = mkOption {
+        type = with types; attrsOf (nullOr (either str (either bool int)));
+        description = ''
+          Trayer configuration as a set of attributes.
```
Comment (sumnerevans) [inline modules/services/trayer.nix:28]: Can you please add a link to the documentation (either a webpage or the manpage)?

File: modules/services/trayer.nix [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+    {
+      home.packages = [ cfg.package ];
+
+      systemd.user.services.trayer = let
+        parameter = let
+          valueToString = v:
+            if isBool v then
+              (if v then "true" else "false")
+            else if (v == null) then
+              "none"
```
Comment (sumnerevans) [inline modules/services/trayer.nix:51]: do we really need to handle this? Can we just filter out `null` values?

File: modules/services/trayer.nix [hunk 6]
```diff
@@ -0,0 +1,85 @@
...
+              "${toString v}";
+        in concatStrings
+        (mapAttrsToList (k: v: "--${k} ${valueToString v} ") cfg.config);
+      in {
+        Unit = {
+          Description = "trayer -- lightweight GTK2+ systray for UNIX desktops";
+          PartOf = [ "tray.target" ];
+        };
+
+        Install = { WantedBy = [ "tray.target" ]; };
```
Comment (sumnerevans) [inline modules/services/trayer.nix:62]: super nit:
```suggestion
        Install.WantedBy = [ "tray.target" ];
```

File: modules/services/trayer.nix [hunk 7]
```diff
@@ -0,0 +1,85 @@
...
+            (if v then "true" else "false")
+          else if (v == null) then
+            "none"
+          else
+            "${toString v}";
+      in concatStrings
+      (mapAttrsToList (k: v: "--${k} ${valueToString v} ") cfg.config);
+    })
+
+  ]);
```
Comment (sumnerevans) [inline modules/services/trayer.nix:84]: why is this necessary when we are already passing everything as a parameter to the service?

File: tests/modules/services/trayer/basic-configuration.nix [hunk 8]
```diff
@@ -0,0 +1,26 @@
...
+        expand = true;
+        monitor = 1;
+      };
+    };
+
+    nmt.script = ''
+      serviceFile=home-files/.config/systemd/user/trayer.service
+
+      assertFileExists $serviceFile
+      assertFileRegex $serviceFile 'ExecStart=.*/bin/trayer --SetDockType true --SetPartialStrut true --edge top --expand true --monitor 1 --padding 6 --tint 0x282c34'
```
Comment (sumnerevans) [inline tests/modules/services/trayer/basic-configuration.nix:23]: we should also assert the file contents of the config file

**Review by sumnerevans — CHANGES_REQUESTED** [review]
Thanks for the PR! Just a few comments.

File: modules/services/trayer.nix (hunk 7 above)
Comment (AndreasMager) [reply modules/services/trayer.nix:86]: Sorry copy & paste left over.

```suggestion
  ]);
```

File: tests/modules/services/trayer/basic-configuration.nix (hunk 8 above)
Comment (AndreasMager) [reply tests/modules/services/trayer/basic-configuration.nix:23]: The config file will be removed.

File: modules/services/trayer.nix (hunk 4 above)
Comment (AndreasMager) [reply modules/services/trayer.nix:28]:
```suggestion
          Trayer configuration as a set of attributes. 
          Details for trayer can be found here: https://github.com/sargon/trayer-srg
          
          edge       <left|right|top|bottom|none> (default:bottom)
          align      <left|right|center>          (default:center)
          margin     <number>                     (default:0)
          widthtype  <request|pixel|percent>      (default:percent)
          width      <number>                     (default:100)
          heighttype <request|pixel>              (default:pixel)
          height     <number>                     (default:26)
          SetDockType     <true|false>            (default:true)
          SetPartialStrut <true|false>            (default:true)
          transparent     <true|false>            (default:false)
          alpha      <number>                     (default:127)
          tint       <int>                        (default:0xFFFFFFFF)
          distance   <number>                     (default:0)
          distancefrom <left|right|top|bottom>    (default:top)
          expand     <false|true>                 (default:true)
          padding    <number>                     (default:0)
          monitor    <number|primary>             (default:0)
          iconspacing <number>                    (default:0)    
```

File: modules/services/trayer.nix (hunk 5 above)
Comment (AndreasMager) [reply modules/services/trayer.nix:74]:
```suggestion
```

**Comment by AndreasMager** [comment]
I have double checked everything.
When i add trayer.nix file to my system config(v21.05) the service don't starts with systemd.

```systemctl --user status trayer.service
● trayer.service - trayer -- lightweight GTK2+ systray for UNIX desktops
     Loaded: loaded (/nix/store/vlczbkcqvlh75473hhk1q4zs2gxr59zr-home-manager-files/.config/systemd/user/trayer.service; enabled; vendor preset: enabled)
     Active: inactive (dead)
```
     
Manually the service starts.
Does anyone have any idea why the service doesn't start.

**Comment by AndreasMager** [comment]
```
sytemctl --user list-dependencies
default.target
● ├─emacs.service
● ├─syncthing.service
● └─basic.target
●   ├─paths.target
●   ├─sockets.target
●   │ ├─dbus.socket
●   │ └─pulseaudio.socket
●   └─timers.target
```
I thing i have found the answer.
The syncthing service doesn't request tray service at the moment.

File: modules/services/trayer.nix [hunk 9]
```diff
@@ -0,0 +1,92 @@
...
+          SetPartialStrut <true|false>            (default:true)
+          transparent     <true|false>            (default:false)
+          alpha      <number>                     (default:127)
+          tint       <int>                        (default:0xFFFFFFFF)
+          distance   <number>                     (default:0)
+          distancefrom <left|right|top|bottom>    (default:top)
+          expand     <false|true>                 (default:true)
+          padding    <number>                     (default:0)
+          monitor    <number|primary>             (default:0)
+          iconspacing <number>                    (default:0)
```
Comment (berbiche) [inline modules/services/trayer.nix:48]: This won't render correctly in the HTML documentation I think.

A good example of doing this is done in the `lf` module:
- https://nix-community.github.io/home-manager/options.html#opt-programs.lf.settings
- https://github.com/nix-community/home-manager/blob/775cb20bd4af7781fbf336fb201df02ee3d544bb/modules/programs/lf.nix#L69-L87

Comment (AndreasMager) [reply modules/services/trayer.nix]: Sorry fixed it

**Comment by berbiche** [comment]
Seems like you need to run the formatting script, `./format` from the project root.

**Comment by AndreasMager** [comment]
@berbiche sorry to late at night...

File: modules/modules.nix [hunk 10]
```diff
@@ -216,6 +216,7 @@ let
     (loadModule ./services/taffybar.nix { })
     (loadModule ./services/tahoe-lafs.nix { })
     (loadModule ./services/taskwarrior-sync.nix { condition = hostPlatform.isLinux; })
+    (loadModule ./services/trayer.nix { })
```
Comment (berbiche) [inline modules/modules.nix:219]: You'll need to rebase your changes because `loadModule` does not exist anymore in `modules/modules.nix`.

File: tests/modules/services/trayer/basic-configuration.nix [hunk 11]
```diff
@@ -0,0 +1,26 @@
+{ config, pkgs, ... }:
+
+{
+  config = {
+    services.trayer = {
+      enable = true;
+      package = pkgs.writeScriptBin "dummy-trayer" "";
```
Comment (berbiche) [inline tests/modules/services/trayer/basic-configuration.nix:7]:
```suggestion
      package = pkgs.writeScriptBin "dummy-trayer" "" // { outPath = "@trayer@"; };
```

File: tests/modules/services/trayer/basic-configuration.nix (hunk 8 above)
Comment (berbiche) [reply tests/modules/services/trayer/basic-configuration.nix:23]:
```suggestion
      assertFileContent $serviceFile <<<'ExecStart=@trayer@/bin/trayer --SetDockType true --SetPartialStrut true --edge top --expand true --monitor 1 --padding 6 --tint 0x282c34'
```

**Comment by rycee** [comment]
Thanks all. I made some additional changes and merged to master. The most notable change is that I renamed the option `services.trayer.config` to `services.trayer.settings` to better correspond to RFC 42. The full changes are

``` nix
diff --git a/modules/misc/news.nix b/modules/misc/news.nix
index 6d3c7f1a..bda956fa 100644
--- a/modules/misc/news.nix
+++ b/modules/misc/news.nix
@@ -2142,7 +2142,8 @@ in
       }
 
       {
-        time = "2021-07-21T22:18:22+02:00";
+        time = "2021-07-23T22:22:31+00:00";
+        condition = hostPlatform.isLinux;
         message = ''
           A new module is available: 'services.trayer'.
         '';
diff --git a/modules/services/trayer.nix b/modules/services/trayer.nix
index ccceb32f..6b5da31c 100644
--- a/modules/services/trayer.nix
+++ b/modules/services/trayer.nix
@@ -9,75 +9,94 @@ let
     values = "true|false";
     default = "true";
   };
+
   number0 = {
     type = types.int;
     values = "number";
     default = "0";
   };
+
   knownSettings = {
     edge = {
       type = types.str;
       values = "left|right|top|bottom|none";
       default = "bottom";
     };
+
     align = {
       type = types.str;
       values = "left|right|center";
       default = "center";
     };
+
     margin = number0;
     widthtype = {
       type = types.str;
       values = "request|pixel|percent";
       default = "percent";
     };
+
     width = {
       type = types.int;
       values = "number";
       default = "100";
     };
+
     heighttype = {
       type = types.str;
       values = "request|pixel";
       default = "pixel";
     };
+
     height = {
       type = types.int;
       values = "number";
       default = "26";
     };
+
     SetDockType = boolTrue;
+
     SetPartialStrut = boolTrue;
+
     transparent = {
       type = types.bool;
       values = "true|false";
       default = "false";
     };
+
     alpha = {
       type = types.int;
       values = "number";
       default = "127";
     };
+
     tint = {
       type = types.str;
       values = "int";
       default = "0xFFFFFFFF";
     };
+
     distance = number0;
+
     distancefrom = {
       type = types.str;
       values = "left|right|top|bottom";
       default = "top";
     };
+
     expand = boolTrue;
+
     padding = number0;
+
     monitor = {
       values = "number|primary";
       type = types.str;
       default = "0";
     };
+
     iconspacing = number0;
   };
+
   cfg = config.services.trayer;
 
 in {
@@ -96,20 +115,21 @@ in {
         description = "The package to use for the trayer binary.";
       };
 
-      config = mkOption {
+      settings = mkOption {
         type = with types; attrsOf (nullOr (either str (either bool int)));
         description = ''
-          Trayer configuration as a set of attributes.
-          Details for trayer can be found here: https://github.com/sargon/trayer-srg
-          <informaltable frame="none">
-          <tgroup cols="4">
-          <tbody>
+          Trayer configuration as a set of attributes. Further details can be
+          found at <link xlink:href="https://github.com/sargon/trayer-srg"/>.
+
+          <informaltable frame="none"><tgroup cols="4">
+          <thead>
           <row>
-          <entry>proprety name</entry>
-          <entry>type</entry>
-          <entry>values</entry>
-          <entry>default</entry>
+          <entry>Property Name</entry>
+          <entry>Type</entry>
+          <entry>Values</entry>
+          <entry>Default</entry>
           </row>
+          </thead><tbody>
           ${concatStringsSep "\n" (mapAttrsToList (n: v: ''
             <row>
               <entry><varname>${n}</varname></entry>
@@ -121,7 +141,6 @@ in {
           </tbody></tgroup></informaltable>
         '';
         default = { };
-        defaultText = literalExample "{ }";
         example = literalExample ''
           {
             edge = "top";
@@ -135,23 +154,28 @@ in {
   };
 
   config = mkIf cfg.enable ({
+    assertions = [
+      (lib.hm.assertions.assertPlatform "services.trayer" pkgs
+        lib.platforms.linux)
+    ];
+
     home.packages = [ cfg.package ];
+
     systemd.user.services.trayer = let
-      parameter = let
-        valueToString = v:
-          if isBool v then (if v then "true" else "false") else "${toString v}";
-      in concatStrings
-      (mapAttrsToList (k: v: "--${k} ${valueToString v} ") cfg.config);
+      valueToString = v:
+        if isBool v then (if v then "true" else "false") else "${toString v}";
+      parameter = k: v: "--${k} ${valueToString v}";
+      parameters = concatStringsSep " " (mapAttrsToList parameter cfg.settings);
     in {
       Unit = {
         Description = "trayer -- lightweight GTK2+ systray for UNIX desktops";
-        PartOf = [ "tray.target" "graphical-session.target" ];
+        PartOf = [ "tray.target" ];
       };
 
-      Install.WantedBy = [ "graphical-session.target" ];
+      Install.WantedBy = [ "tray.target" ];
 
       Service = {
-        ExecStart = "${cfg.package}/bin/trayer ${parameter}";
+        ExecStart = "${cfg.package}/bin/trayer ${parameters}";
         Restart = "on-failure";
       };
     };
diff --git a/tests/modules/services/trayer/basic-configuration.nix b/tests/modules/services/trayer/basic-configuration.nix
index 4f096049..23632331 100644
--- a/tests/modules/services/trayer/basic-configuration.nix
+++ b/tests/modules/services/trayer/basic-configuration.nix
@@ -7,7 +7,7 @@
       package = pkgs.writeScriptBin "dummy-trayer" "" // {
         outPath = "@trayer@";
       };
-      config = {
+      settings = {
         edge = "top";
         padding = 6;
         SetDockType = true;
@@ -22,7 +22,8 @@
       serviceFile=home-files/.config/systemd/user/trayer.service
 
       assertFileExists $serviceFile
-      assertFileRegex $serviceFile 'ExecStart=.*/bin/trayer --SetDockType true --SetPartialStrut true --edge top --expand true --monitor 1 --padding 6 --tint 0x282c34'
+      assertFileContains $serviceFile \
+        'ExecStart=@trayer@/bin/trayer --SetDockType true --SetPartialStrut true --edge top --expand true --monitor 1 --padding 6 --tint 0x282c34'
     '';
   };
 }
```
